### PR TITLE
Migrate <Modal>

### PR DESCRIPTION
**Pull request by garfieldduck**
Asana: https://app.asana.com/0/347798529122928/550711635750738

### Purpose
Bring back `<Modal>` from ic-framework-react and ic-framework .

### Implementations
1. Add `<Modal>` which supports the `size` modifier
2. Add the `closable` HOC outside the `gyp-modal__container` in order to make the modal closable by clicking the visible area of `<Backdrop>`. 
3. Add the `className` props into the HOC `closable` to make the styles of `closable` adjustable. (It also makes the closable area inside the modal fit the outer container)
4. Add examples containing modals of different sizes and a closable modal with the header into the storybook.

### Demo
screen shot 2018-02-09 at 4 34 18 pm

### CODE REVIEW

File: packages/core/src/Modal.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+function Modal(props) {
+    const {
+        // React props
+        className,
+        // Other props
+        size,
+    } = props;
+    const rootClassName = classNames(BEM.root.toString(),
+                                     className, `${BEM.root.modifier(size)}`);
```
Comment (zhusee2) [inline packages/core/src/Modal.js]: icBEM 本身可以 chain，所以這邊寫成這樣就可以了

```js
// bemClass.toString() ==> 'gyp-modal gyp-modal--small'
const bemClass = BEM.root.modifier(size);
const rootClassName = className(bemClass.toString(), className);
```

另外需要換行時，請參考這邊的規則：
https://github.com/airbnb/javascript#whitespace--max-len
https://github.com/airbnb/javascript#commas--dangling

```js
invokeFunction(
    param1,
    param2,
    param3,
);
```

File: packages/core/src/Modal.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    onEscape: true,
+    onClickOutside: true,
+    onClickInside: false,
+})(ModalContent);
+
+
+function Modal(props) {
+    const {
+        // React props
+        className,
```
Comment (zhusee2) [inline packages/core/src/Modal.js]: 直接用 object destructing 使用 props
(Ref: https://github.com/airbnb/javascript#destructuring--object)

```js
function Modal(
  size,
  // React props
  className,
) {}
```
排序部分純粹是這邊的習慣，盡量把 component 自己定義的 props 放上面、因為別人而來的 props 放下面、React 的放最下面。

File: packages/core/src/Modal.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    children,
+    // Other props
+    header,
+    bodyClassName,
+    bodyPadding,
+}) => (
+    <div className={BEM.container}>
+        {renderHeader(header, `${BEM.header}`)}
+        <div
+            className={classNames(
```
[the remark this answers is not in the document]
Comment (garfieldduck) [reply packages/core/src/Modal.js]: 👍

**Comment by garfieldduck** [comment]
@zhusee2: The test for <Modal> has been added. You may take a look if you are free 😄

File: packages/core/src/Modal.js [hunk 4]
```diff
@@ -86,13 +85,12 @@ const ClosableModalContent = closable({
...
+        size,
         // React props
         className,
-        // Other props
-        size,
     } = props;
-    const rootClassName = classNames(BEM.root.toString(),
-                                     className, `${BEM.root.modifier(size)}`);
+    const bemClass = BEM.root.modifier(size);
+    const rootClassName = className(bemClass.toString(), className);
```
Comment (zhusee2) [inline packages/core/src/Modal.js]: 這邊拼錯了 `className**s**` 所以你 test 跑不過

File: packages/core/src/Modal.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+};
+
+const ClosableModalContent = closable({
+    onEscape: true,
+    onClickOutside: true,
+    onClickInside: false,
+})(ModalContent);
+
+
+function Modal(props) {
```
Comment (cjies) [inline packages/core/src/Modal.js]: As the recommendation from @zhusee2, destructure props at here XD

```js
function Modal({ size, className }) {
 ...
}
```

Comment (garfieldduck) [reply packages/core/src/Modal.js]: Thanks for reminding me :)

File: packages/core/src/Modal.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhusee2) [reply packages/core/src/Modal.js]: Please don't forget here 👀

File: packages/core/src/Modal.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    // React props
+    className,
+}) {
+    const bemClass = BEM.root.modifier(size);
+    const rootClassName = classNames(bemClass.toString(), className);
+
+    return (
+        <article className={rootClassName}>
+            <Overlay />
+            <ClosableModalContent
```
Comment (zhusee2) [inline packages/core/src/Modal.js]: I suggest to forward `closable` prop to `<ClosableModalContent>` here from `<Modal>`, so you can easily override `closable()` configs.

File: packages/core/src/__tests__/Modal.test.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            expect(wrapper.hasClass(
+                BEM.root
+                    .modifier(size)
+                    .toString({ stripBlock: true })
+            )).toBeTruthy();
+        };
+
+        testSizeProp('small');
+        testSizeProp('large');
+        testSizeProp('full');
```
Comment (zhusee2) [inline packages/core/src/__tests__/Modal.test.js]: 👍

File: packages/core/src/Modal.js (hunk 6 above)
Comment (garfieldduck) [reply packages/core/src/Modal.js]: Thanks so much! I've added the closable prop back.

**Pull request merged**